partiql-eval/src/eval/mod.rs

```diff
@@ -140,33 +144,52 @@ pub struct Evaluated {
 }
 
 /// Represents an evaluation context that is used during evaluation of a plan.
-pub trait EvalContext {
-    fn bindings(&self) -> &dyn Bindings<Value>;
+pub trait EvalContext<'a>: SessionContext<'a> + Debug {
+    fn as_session(&'a self) -> &'a dyn SessionContext<'a>;
+
     fn add_error(&self, error: EvaluationError);
     fn has_errors(&self) -> bool;
     fn errors(&self) -> Vec<EvaluationError>;
 }
 
-#[derive(Default, Debug)]
-pub struct BasicContext {
-    bindings: MapBindings<Value>,
-    errors: RefCell<Vec<EvaluationError>>,
+#[derive(Debug)]
+pub struct BasicContext<'a> {
+    pub bindings: MapBindings<Value>,
+
+    pub sys: SystemContext,
+    pub user: HashMap<String, &'a (dyn Any)>, // TODO: Unicase the keys?
+
+    pub errors: RefCell<Vec<EvaluationError>>,
 }
 
-impl BasicContext {
-    pub fn new(bindings: MapBindings<Value>) -> Self {
+impl<'a> BasicContext<'a> {
+    pub fn new(bindings: MapBindings<Value>, sys: SystemContext) -> Self {
         BasicContext {
             bindings,
+            sys,
+            user: Default::default(),
             errors: RefCell::new(vec![]),
         }
     }
 }
-
-impl EvalContext for BasicContext {
+impl<'a> SessionContext<'a> for BasicContext<'a> {
     fn bindings(&self) -> &dyn Bindings<Value> {
         &self.bindings
     }
 
+    fn system_context(&self) -> &SystemContext {
+        &self.sys
+    }
+
+    fn user_context(&self, name: &str) -> Option<&(dyn Any)> {
```

Comment:
Is there _any_ way this API can output a generic instead of a `Option<&dyn Any>`?  I'd really like to implement these traits myself and just return a type that I choose, instead of going through a `dyn Any`.

Reply:
If there is, I don't see how.

Making it a generic parameter of the `SessionContext` would mean that any and all extensions that are loaded would need to be able to deal with that single type.